utilities/gitopsctl/implementations/parse-json-logs/parse-json-logs.go

```diff
@@ -0,0 +1,691 @@
+	separatorTextColor       = color.New(color.FgMagenta).Add(color.Bold).SprintFunc()
+)
+
+func ParseJsonLogsFromStdin() {
+
+	lineCount := 1
+
+	scanner := bufio.NewScanner(os.Stdin)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		parseJSONLogLine(line, lineCount)
+
+		lineCount++
+	}
+
+	if scanner.Err() != nil {
+		fmt.Println(scanner.Err().Error())
+		os.Exit(1)
+		return
+	}
+}
+
+func ReadAllLinesFirstThenSortByTimestamp() {
+
+	lines := []string{}
+
+	scanner := bufio.NewScanner(os.Stdin)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		lines = append(lines, line)
+	}
+
+	if scanner.Err() != nil {
+		fmt.Println(scanner.Err().Error())
+		os.Exit(1)
+		return
+	}
+
+	// Extract timestamp and sort by that
+	sort.Sort(ByTS(lines))
+
+	// Parse in sorted order
+	for _, line := range lines {
+		parseJSONLogLine(line, 0)
+	}
+
+}
+
+func parseJSONLogLine(line string, lineNumber int) {
+
+	// A) If the log line is from a Goreman log, pre-format it to make it parseable by JSON
+	isGoreman, jsonModified, line, beforeJsonLocalDevPrefix := parseGoremanLogsIfApplicable(line)
+
+	if isGoreman && !jsonModified {
+		// We weren't able to extract any JSON, so just print the line as is
+		fmt.Println(line)
+		return
+	}
+
+	fullJsonMap := map[string]any{}
+
+	if err := json.Unmarshal(([]byte)(line), &fullJsonMap); err != nil {
+		// We weren't able to extract any JSON, so just print the line as is
+		fmt.Println(line)
+		return
+	}
+
+	var structuredMap map[string]any
+
+	var output string
+
+	structuredMapVal, exists := fullJsonMap["structured"]
+	if exists {
+
+		// B) If the 'structured' field exists, then the log is from splunk
+
+		structuredMap = (structuredMapVal).(map[string]any)
+
+		delete(fullJsonMap, "structured")
+
+		output = parseJSONMapFromLine(structuredMap, fullJsonMap, lineNumber)
+	} else {
+
+		_, hasTimestampField := fullJsonMap["@timestamp"]
+
+		if hasTimestampField {
+
+			// C) If the 'structured' field doesn't exist and @timestamp DOES exist, then the log is direct splunk output
+			output = parseJSONMapFromLine(make(map[string]any), fullJsonMap, lineNumber)
+
+		} else {
+			// D) If the 'structured' field doesn't exist and @timestamp doesn't exist, then the log is direct controller output (e.g. not from splunk)
+			output = parseJSONMapFromLine(fullJsonMap, make(map[string]any), lineNumber)
+		}
+
+	}
+
+	// Output the line (optionally with prefix for the Goreman case, only)
+
+	fmt.Println(beforeJsonLocalDevPrefix + output)
+	fmt.Println()
+
+}
+
+// The Goreman log lines need to be carefully cut up to allow us to both parse the JSON, but also preserve the Goreman color prefix.
+func parseGoremanLogsIfApplicable(line string) (bool, bool, string, string) {
+	var beforeJsonLocalDevPrefix string
+
+	isGoremanLocalDev := false // true if the logs are from goreman, false otherwise
+
+	goremanPrefixes := []string{
+		" backend | ", " appstudio-controller | ", " cluster-agent | ",
```

Comment:
I wonder if this parsing can be made more generic to use other logs, like Argo CD

Reply:
True, if someone wanted to tackle this as part of another PR that would be :+1: with me.